<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Manual</title>
  <style>
    body {
      margin: 0;
      padding: 0;
      font-family: Arial, sans-serif;
      line-height: 1.6;
      color: #222;
      background: #fff;
    }
    main { max-width: 900px; margin: 2rem auto; padding: 0 1rem; }
    h1, h2, h3, h4, h5, h6 { margin-top: 1.8em; margin-bottom: 0.6em; line-height: 1.25; }
    h1 {text-align: center;}
    h2, h3 {color: #002060;}
    h4 {color: #800080;}
    h5 {font-weight: bold;}
    p { margin: 0.8em 0; }
    ul, ol { padding-left: 1.5em; }
    code, pre { font-family: Consolas, "Courier New", monospace; font-size: 1em; }
    pre { white-space: pre-wrap; background: #f7f7f7; border: 1px solid #e0e0e0; border-radius: 4px; padding: 0.75rem; overflow-x: auto; }
    a { color: #0645ad; text-decoration: none; }
    a:hover { text-decoration: underline; }
    hr { border: 0; border-top: 1px solid #ddd; margin: 2rem 0; }
    table { border-collapse: collapse; }
    th, td { border: 1px solid #ddd; padding: 0.5rem; }
  </style>
</head>
<body>
  <main>
  <h1 id="strainvis-a-python-based-web-application-for-interactive-visual-analysis-of-strain-tracking-methods">StrainVis: a Python-based web application for interactive visual analysis of strain-tracking methods</h1>
<h4 id="version-1.0.0">Version 1.0.0</h4>
<ul>
<li><a href="#overview">Overview</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#open-strainvis-web-application">Open StrainVis web-application</a></li>
  <ul>
    <li><a href="#open-several-strainvis-sessions">Open several StrainVis sessions</a></li>
  </ul>
<li><a href="#run-strainvis-on-a-remote-server-and-open-it-in-a-local-browser">Run StrainVis on a remote server and open it in a local browser</a></li>
<li><a href="#input">Input</a></li>
<li><a href="#visual-analyses">Visual analyses</a></li>
  <ul>
    <li><a href="#customizing-the-plots">Customizing the plots</a></li>
    <li><a href="#saving-the-plots">Saving the plots</a></li>
  </ul>
<li><a href="#single-species-analyses">Single species analyses</a>
<ul>
<li><a href="#syntracker-results-single-species-analyses">Single-species analyses of SynTracker results</a>
<ul>
<li><a href="#apss-based-analyses">APSS-based analyses</a>
<ul>
<li><a href="#initial-bar-plots">Initial bar-plots</a></li>
<li><a href="#apss-distribution-plot">APSS distribution plot</a></li>
<li><a href="#clustered-heatmap-plot">Clustered heatmap plot</a></li>
<li><a href="#network-plot">Network plot</a></li>
</ul></li>
<li><a href="#synteny-per-position-analyses">Synteny per position analyses</a></li>
</ul></li>
<li><a href="#ani-results-analyses">Single-species analyses of ANI results</a></li>
<li><a href="#combined-analysis">Single-species combined analysis</a></li>
</ul></li>
<li><a href="#multiple-species-analyses">Multiple species analyses</a>
<ul>
<li><a href="#setting-the-included-species">Setting the included species</a></li>
<li><a href="#syntracker-results-multi-species-analysis">Multiple-species analysis of SynTracker results</a>
<ul>
<li><a href="#initial-bar-plots">Initial bar-plots</a></li>
<li><a href="#apss-distribution-among-species-plot">APSS distribution among species plot</a></li>
</ul></li>
<li><a href="#ani-multi-species-analysis">Multiple-species analysis of ANI results</a>
<ul>
<li><a href="#ani-distribution-among-species-plot">ANI distribution among species plot</a></li>
</ul></li>
</ul></li>
</ul>
<h2 id="overview">Overview</h2>
<p>StrainVis is a Python-based web application for visual analyses and interactive exploration of the results obtained by
the SynTracker pipeline or by other strain tracking methods, based on ANI (Average Nucleotide Identity).
</p>
<p>StrainVis accepts either SynTracker’s output file ‘synteny_scores_per_region.csv’ or an ANI file, obtained by another method, containing either one reference genome or multiple reference genomes (usually, one reference genome per species). It presents accordingly analyses for each species separately and for multiple species together.<br />
A metadata file (matches to the samples that were previously compared by SynTracker) may also be provided in order to enable a deeper analysis of SynTracker’s results.</p>
<p>StrainVis allows the user to interactively select the presented plots and to change visual parameters in each plot. It also enables to interactively select the metadata feature by which the samples should be grouped and coloured (for each plot separately).</p>
<p>Each of the presented plots can be downloaded and saved as a high-resolution image, in several common image formats.</p>
<h2 id="installation">Installation</h2>
<p>Download the latest release of StrainVis from: <a href="https://github.com/leylabmpi/StrainVis/releases" target="_blank">https://github.com/leylabmpi/StrainVis/releases</a>.</p>
<p>Extract the tar.gz file into the desired working-directory.</p>
<p>Create a new conda environment using the ‘StrainVis.yml’ file (located in the root directory of SynTrakcerVis) using the following command:<br />
<code>conda env create -f StrainVis.yml</code></p>
<p>Activate the newly created environment: <code>conda activate StrainVis</code></p>
<h2 id="open-strainvis-web-application">Open StrainVis web-application</h2>
<p>StrainVis is built using <a href="https://panel.holoviz.org/index.html" target="_blank">Panel - a powerful data exploration &amp; web app framework for Python</a>. The web-application is served using the Bokeh server. To launch the server from the command-line and open the application in the browser, type the following command (from within the activated conda strain_vis environment):</p>
<p><code>panel serve strain_vis.py --port 5005 --websocket-max-message-size 524288000 &amp;</code></p>
<p>The application should be opened in the browser under the URL: http://localhost:5005/strain_vis.<br />
As long as the Bokeh server is running, StrainVis application can be accessed using the above URL.<br />
Please note that several instances of StrainVis can be opened simultaneously in different browser windows/tabs.<br />
It is also possible to launch more than several server processes simultaneously using different ports.</p>
<p><strong>Stop the server</strong>: In order to stop the Bokeh server, its running process should be killed.</p>
<h3 id="open-several-strainvis-sessions">Open several StrainVis sessions</h3>
<ul>
<li><strong>Several user-sessions using the same web-server instance:</strong> It is possible to open as many user sessions as needed under the same server instance that was started using the ‘panel serve…’ command. All sessions will be accessible under the same URL (for example: http://localhost:5005/strain_vis). Each session works separately and can process a different dataset, but since all of them are executed by the same process, if one session is busy processing a heavy-duty task, it will affect all the other user sessions.</li>
<li><strong>Several web-server instances:</strong> In order to process different datasets at the same time using separated computational resources, it is needed to start several web-server instances listening to different ports. It simply means to run the ‘panel serve…’ command using a different port number for each StrainVis instance. Each web-server instance will be accessible under: http://localhost:[port_number]/strain_vis.</li>
</ul>
<h2 id="run-strainvis-on-a-remote-server-and-open-it-in-a-local-browser">Run StrainVis on a remote server and open it in a local browser</h2>
<p>It is possible to run StrainVis on a remote server or cloud service in cases of large datasets, in order to improve the performance. This can be done according to the following steps:</p>
<ol type="1">
<li><p><strong>Installing StrainVis:</strong> StrainVis shoul be installed on the remote server as explained above.</p></li>
<li><p><strong>Open SSH-tunnel:</strong> Create an SSH tunnel from your local machine to the remote server:<br />
<code>ssh -L 5006:localhost:5006 user@remote-server</code></p></li>
<li><p><strong>Start the Bokeh server on the remote server:</strong><br />
From within the activated conda strain_vis environment, run the following command:<br />
<code>panel serve strain_vis.py --port 5006 --websocket-max-message-size 524288000 &amp;</code><br />
Note that it is important not to use the –show option when SynTracker runs on a remote machine.</p></li>
<li><p><strong>Open the application in the local browser:</strong> StrainVis should be accessible under: http://localhost:5006/strain_vis .</p></li>
</ol>
<h2 id="input">Input</h2>
<h3 id="mandatory-input">Mandatory input:</h3>
<p>StrainVis accepts two types of input files:</p>
    <ul>
      <li>SynTracker’s output file ‘synteny_scores_per_region.csv’ for one or multiple species.</li>
      <li>
        A tab-delimited ANI file for one or multiple species in the following format:<br />
        ‘Ref_genome’, ‘Sample1’, ‘Sample2’, ‘ANI’
      </li>
    </ul>
<p>The user should select one of the following three modes of execution, depending on the available input data:</p>
    <ul>
      <li><p><strong>SynTracker output file</strong>: Analyse the results obtained by the SynTracker pipeline only.</p></li>
      <li><p><strong>ANI file</strong>: Analyse the ANI results obtained by another strain-tracking method.</p></li>
      <li>
        <p><strong>Both SynTracker and ANI files</strong>: Analyse both types of input data (each one separately and combined).<br />
        In this case, the names of the analysed species and the sample IDs must be identical between the two input files.</p>
      </li>
    </ul>
<p>Note that if the input files are bigger than 300 Mb, they cannot be selected via the FileInput widget, but their full path should be typed into the TextInput field.</p>
<h3 id="optional-input">Optional input:</h3>
<p>A metadata file in tab-delimited format. The first column must contain the sample IDs, that should match the sample IDs in the uploaded input file(s). The metadata file may contain an unlimited number of columns (features).</p>
<h2 id="visual-analyses">Visual analyses</h2>
<p>When uploading an input file containing more than one species, StrainVis presents both single species analyses and multiple species analyses in separate tabs.</p>
<ul>
<li><p><strong>Single species analyses</strong>: The analyses are performed on one species at a time. The species can be selected from a drop-down menu, containing all the species in the input file. The list of species may be sorted by the number of compared sample-pairs (in order to display the more abundant species first), or by the species names.</p></li>
<li><p><strong>Multiple species analyses</strong>: The analyses are performed on all the species together or for a selected subset.</p></li>
</ul>
<h3 id="customizing-the-plots">Customizing the plots</h3>
<p>Each plot allows the user to interactively set/change some visual parameters, like color, colormap, etc. In some of the plots it is possible to show/hide elements and to set other parameters which influence the data visualization.</p>
    <p><strong>Including metadata in the plots:</strong></p>
<p>When a metadata file is uploaded, the metadata can be incorporated into the plots. The user can interactively select a feature, by which the presented data will be colored, grouped or filtered.</p>
<h3 id="saving-the-plots">Saving the plots</h3>
<p>The plots can be saved either as images or as text files, containing the underlying data of the plots.<br />
The user may enter the name of the file (including full path), or use the default name and path provided by StrainVis (by default the files are saved under the ‘StrainVis/Downloads/’ directory).</p>
<ul>
<li><p><strong>Save as image:</strong> Each one of the plots can be saved as a high-resolution image in one of the following formats: png, pdf, svg, eps.</p></li>
<li><p><strong>Save data table:</strong> The data table that was used to create the plot can be saved as text in a delimited format.</p></li>
</ul>
<h2 id="single-species-analyses">Single species analyses</h2>
<p>When providing both SynTracker and ANI results, it is possible to move between three available views:</p>
    <ul>
      <li><p><strong>SynTracker</strong>: Presents the single-species analyses of the synteny scores.</p></li>
      <li><p><strong>ANI</strong>: Presents the single-species analyses of the ANI results.</p></li>
      <li><p><strong>Combined</strong>: Presents a combined APSS and ANI based analysis.</p></li>
    </ul>
<h2 id="syntracker-results-single-species-analyses">Single-species analyses of SynTracker results</h2>
<p>StrainVis provides two types of analyses of the synteny scores obtained by SynTracker:</p>
<ol type="1">
<li><p><strong>APSS-based analyses:</strong> Analyses based on the APSS (Average Pairwise Synteny Scores) that are calculated using N sub-sampled regions (according to the user’s choice).</p></li>
<li><p><strong>Synteny per position:</strong> Presenting different measures based on the synteny scores for each position in the reference genome.</p></li>
</ol>
<h3 id="apss-based-analyses">APSS-based analyses</h3>
<h4 id="initial-bar-plots">Initial bar-plots</h4>
<p>The initial presented bar-plots allow the user to interactively change the number of sub-sampled regions and see how it influences the number of sample-pairs comparisons and the total number of samples which can be taken into account in the following APSS-based analyses.<br />
The numbers of sub-sampled regions available for selection are: 40, 60, 80, 100, 125, 150, 175, 200, 250, 300, 350, 400. In case the number of compared pairs, when selecting 40 regions, is lower than 100, the option ‘All regions’ is also available for selection.<br />
Setting the desired subsampling number is done using the slider widget. That means that all the compared pairs from all the available regions are included in the downstream analyses.</p>
<p>By clicking the ‘Display plots using the selected number of regions’ button, the following downstream analyses are being calculated and displayed.</p>
<h4 id="apss-distribution-plot">APSS distribution plot</h4>
<p>This plot shows the APSS distribution among all the compared sample-pairs.<br />
The plot type can be changed between scatter (jitter) plot and boxplot, so as the plot color(s).</p>
    <p><strong>Including metadata:</strong></p>
<p>Upon feature selection, the comparisons can be divided into two categories: same / different feature.<br />
for example: if the selected feature is ‘country’, the two categories are ‘same country’ and ‘different country’.<br />
The features are derived from the upoaded metadata file and can be interactively selected from a drop-down menu.</p>
<h4 id="clustered-heatmap-plot">Clustered heatmap plot</h4>
<p>This plot presents the APSS (Average Pairwise Synteny Scores) of the included pairwise comparisons as a clustered heatmap. The colormap for the scores can be interactively selected from a list of available colormaps.</p>
<p><strong>Including metadata:</strong></p>
    <p>An additional column, coloring the rows by a requested metadata feature, can be added when checking the ‘Use metadata for coloring’ option.<br />
      <ul>
        <li>Color rows by: select a metadata feature, by which the rows in the additional column will be colored.</li>
        <li>
          Select colormap: select a colormap from the drop-down menu to color the different groups of the selected feature.<br />
          When selecting the ‘Define custom colormap’ option, the ‘Custom colormap’ text input widget becomes active.
        </li>
        <li>
          Custom colormap: here the user can enter a list of colors, separated by commas.
          The colors can be provided as standard names (like: red, blue) or Hex-RGB values (like: #FF0000).<br />
          A detailed guide for color notations can be found : <a href="https://www.w3.org/TR/css-color-4/#named-colors" target="_blank">here</a>.</li>
      </ul>
<p>Please note that in case the number of samples exceeds the limit of 150, the heatmap plot cannot be displayed properly. The scoring matrix is provided for download and can be used in another visualization program.</p>
<h4 id="network-plot">Network plot</h4>
<p>This plot presents the samples as a 2D network using the Fruchterman-Reingold force-directed graph layout. The layout algorithm clusters the nodes (in this case, the samples) iteratively, using the APSS as the weight attribute for the network. The higher the APSS between two samples, the closer the two nodes will be positioned in space.<br />
The APSS threshold (by default, the mean APSS) determines whether two samples are connected in the network or not. This influences the clustering process of the network, as connected nodes are being clustered closer together than unconnected nodes.<br />
Please note that if the number of samples exceeds the limit of 300, the network plot cannot be well displayed including all its interactive features. It ia possible to download the network data in tsv format and visualize it using another program.</p>
<p><strong>Interaction with the network plot:</strong></p>
<p>The network plot is created using the Bokeh backend. The Bokeh interface allows the following interactions with the plot using the mouse:</p>
    <ul>
      <li>Panning (dragging and moving the plot)</li>
      <li>Zoom-in / zoom-out</li>
      <li>Select a specific area and display it</li>
      <li>Reset to initial display</li>
      <li>Hover tool: displays the relevant information about each sample when hovering the network nodes.</li>
    </ul>
<p><strong>Customization options of the network (unrelated to metadata):</strong></p>
<ul>
<li><strong>Threshold for network connections:</strong> Enables to set the APSS threshold to define whether two nodes (samples) are connected in the network or not (If the APSS of a comparison between two samples is equals to or greater than the threshold, the samples are connected in the network). The threshold affects the clustering of the network when increasing the number of iterations. A higher threshold means less connections between the nodes, which results in a more fragmented network, composed of a larger number of smaller clusters.<br />
StrainVis provides three pre-defined threshold options:<br />
Mean APSS (among all pairwise comparisons)<br />
Mean APSS + 1 std<br />
Mean APSS + 2 std (when the value &lt;= 0.99)<br />
By selecting the ‘Define another threshold’ option, it is possible to set a different threshold (between 0.5 and 1.0) using the ‘Define threshold’ widget.</li>
<li><strong>Number of iterations:</strong> Number of clustering iterations of the network, performed using the Fruchterman-Reingold algorithm. With each iteration, the nodes that have higher APSS score, become closer to each other in the 2D space and the network becomes more clustered. The slider widget allows to select the number of iterations between 50 and 500.</li>
<li><strong>Initialize nodes positions:</strong> Clicking this button assigns the nodes new initial positions in the 2D space (by random) and starts the clustering process from the beginning, performing the selected number of iterations.</li>
<li><strong>Nodes / Edges color:</strong> Enable to set a unified color for the nodes or for the edges of the network, using a color-picker widget.</li>
<li><strong>Show sample names:</strong> When checked, the sample names are displayed on top of the network nodes.</li>
</ul>
<p><strong>Customization options when metadata is provided:</strong></p>
<ul>
<li><strong>Color nodes by:</strong> Select a metadata feature by which the nodes will be grouped and colored. A legend for the colored groups is added to the network plot (in case the number of groups does not exceed 10).</li>
<li><strong>Continuous feature:</strong> If the selected feature has numeric continuous values (for example: age), this checkbox should be checked. The variety of colormaps available for selection changes accordingly and a colorbar is added to the network plot.<br />
Note that only features which numerical values can be checked as continuous.</li>
<li><strong>Select colormap for nodes:</strong> Select a colormap from the drop-down menu to color the nodes by the different groups of the selected metadata feature. A different set of colormaps is provided for categorical data and for continuous data.</li>
<li><strong>Define custom colormap:</strong> For categorical features, it is also possible to define a custom list of colors for the different groups. This option becomes active when selecting the ‘Define custom colormap’ option from the colormaps drop-down menu. A list of colors, separated by commas, can be entered to the text-input widget. The colors can be provided as standard names (like: red, blue) or Hex-RGB values (like: #FF0000).<br />
A detailed guide for color notations can be found : <a href="https://www.w3.org/TR/css-color-4/#named-colors" target="_blank">here</a>.</li>
<li><strong>Color edges by feature (same/different):</strong> Checking this option enables to select a feature, by which the edges (connections) are divided into two categories and can be colored differently. One category is the connections between samples that belong to the same group and the other is the connections between samples that belong to a different group (of the selected feature). For example, if the selected feature is ‘country’, the edges between nodes that belong to the same country can be colored differently than the edges between the nodes that belong to a different country.</li>
<li><strong>Color edges by:</strong> Select the feature, by which the edges are divided into two categories.</li>
<li><strong>Same color / Different color:</strong> Select the color that will be applied on each category of edges.</li>
</ul>
<p><strong>Saving options:</strong></p>
<ul>
<li><strong>Save as image:</strong> When using the ‘Plot download options’ and selecting one of the four available image formats, the network graph area is saved in its initial presentation (without considering interactions such as zoom-in/out, etc.).<br />
In order to save the exact current view (including graph interactive modifications), it is possible to use the Bokeh interface ‘Save’ button, which exports the graph in png format only (it is then saved in the browser’s default Downloads location).</li>
<li><strong>Save data table:</strong> The network data is saved as a tab-delimited format file, containing the following columns: Sample1, Sample2, APSS, weight.<br />
The weight column reflects the score of each pair of samples after applying the APSS connections threshold.</li>
</ul>
<h3 id="synteny-per-position-analyses">Synteny per position analyses</h3>
<p>These analyses are presented for each contig of the selected reference genome. If there is more than one contig, it can be selected from the list of contigs using a drop-down menu. By default, the contigs are sorted by their length, but they can be sorted by their names as well.</p>
<p>The synteny per position plot displays the following four analyses on top of each other, where each analysis can be marked as shown or hidden in the plot.</p>
<ol type="1">
<li><p><strong>Average synteny scores per region:</strong> This line-plot shows the average synteny score for each region in the reference genome. It is calculated based on all the sample-pairs comparisons derived from a specific region.</p></li>
<li><p><strong>All synteny scores per region:</strong> This horizontal line plot shows all the synteny scores for each region on the reference genome.</p></li>
<li><p><strong>Hypervariable regions:</strong> Highlight regions that meet the following criteria:<br />
They are among the bottom 10% regions with the lowest average synteny scores.<br />
They appear in at least 10% of the compared sample-pairs.</p></li>
<li><p><strong>Hyperconserved regions:</strong> Highlight regions that meet the following criteria:<br />
They are among the top 10% regions with the highest average synteny scores.<br />
They appear in at least 50% of the compared sample-pairs.</p></li>
</ol>
<p><strong>Customization options (unrelated to metadata):</strong></p>
<ul>
<li><strong>Set contig length range:</strong> Set the start and end positions of the contig that will be presented in the plot.</li>
<li><strong>Set new range button:</strong> Clicking this button updates the plot to present the newly set range.</li>
<li><strong>Reset range button:</strong> Clicking this button resets the range to show the whole contig in the plot.</li>
<li><strong>Colors:</strong> The color of each one of the plots can be set separately, as well as the alpha transparency of the highlighted conserved / variable regions.</li>
</ul>
<p><strong>Customization options when metadata is provided:</strong></p>
<ul>
<li><strong>Filter plot by metadata:</strong> The metadata feature, by which the presented data will be filtered, can be selected from the drop-down menu.</li>
<li><strong>Include the following groups in the plot:</strong> It is possible to select one or more groups to be included in the plot.</li>
<li><strong>Filter plot button:</strong> Clicking this button updates the plot, so that only pairwise comparisons, originating from the selected groups of the selected feature, will be included in the plot.</li>
<li><strong>Reset filteration button:</strong> Clicking this button resets the filtering and updates the plot so that all data is shown.</li>
</ul>
<h3 id="ani-results-analyses">Single-species analyses of ANI results</h3>
<p>The following analyses are provided for the ANI results:</p>
    <ul>
      <li><p><div style="color: #800080; font-weight: bold;">ANI distribution plot:</div> Shows the ANI distribution among all the sample pairs.</p></li>
      <li><p><div style="color: #800080; font-weight: bold;">Clustered heatmap plot:</div> Presents the ANI of the pairwise comparisons as a clustered heatmap.</p></li>
    </ul>
<p>These plots are similar to the ones created for the SynTracker results.
    Detailed information about the customization options can be found under the sections <a href="#apss-distribution-plot">APSS distribution plot</a>
 and <a href="#clustered-heatmap-plot">Clustered heatmap plot</a> of the SynTracker results.</p>
<h3 id="combined-analysis">Single-species combined analysis</h3>
<p>This tab is available when the user provides both SynTracker and ANI input files.</p>
<h4 id="ani-vs.-apss-scatter-plot">ANI vs. APSS scatter plot</h4>
<p>This plot is a scatter plot of the ANI vs. the APSS among all the sample pairwise comparisons, which are available in both input files. The APSS values that are shown in the plot correspond to the number of subsampled regions that was selected by the user in the ‘APSS-based analyses’ view.<br />
The Spearman correlation between the two metrices is calculated and presented on the plot.<br />
The color of the data points can be set by the user.</p>
<p><strong>Including metadata</strong>:</p>
    <p>Upon feature selection, the data points, representing the pairwise-comparisons, can be colored according to two categories: same / different feature. for example: if the selected feature is ‘country’, the two categories are ‘same country’ and ‘different country’. The features are derived from the upoaded metadata file and can be interactively selected from a drop-down menu. The ‘same color’ and the ‘different color’ can be set by the user.</p>
<h2 id="multiple-species-analyses">Multiple Species Analyses</h2>
<p>The multiple species analyses tab is active when the input file(s) contains more than one species.
    It presents APSS-/ANI-based analysis for all the species or for a selected subset of species.<br />
    Please note that when both SynTracker and ANI input files are provided, the list of available species is taken from the SynTracker results input file.
</p>
<h3 id="setting-the-included-species">Setting the included species</h3>
<ul>
    <li><p><strong>All species:</strong> Include all available species.</p></li>
    <li><p><strong>Select a subset of species:</strong> Using the multi-select widget, it is possible to select specific species to be included in the analysis.</p></li>
    <li><p><strong>Sort species by:</strong> The list of species may be sorted by the number of compared sample-pairs (in order to display the more abundant species first), or alphabetically, by the species names.</p></li>
    <li><strong>Update species selection/sorting button:</strong> Clicking this button updates the plot with the selected set of species and their selected order.</li>
</ul>
<h3 id="syntracker-results-multi-species-analysis">Multiple-species analysis of SynTracker results</h3>
<h4 id="initial-bar-plots-1">Initial bar-plots</h4>
<p>The initial presented bar-plots allow the user to interactively change the number of sub-sampled regions and see how it influences the number of compared sample-pairs and the total number of species that can be taken into account in the following APSS-based analysis.<br />
The numbers of sub-sampled regions available for selection are: 40, 60, 80, 100, 125, 150, 175, 200, 250, 300, 350, 400. In case the number of compared pairs, when selecting 40 regions, is lower than 100, the option ‘All regions’ is also available for selection. That means that all the compared pairs from all the available regions are included in the downstream analysis.<br />
Setting the desired subsampling number is done using the slider widget.</p>
<p>By clicking the ‘Display plots using the selected number of regions’ button, the following downstream analyses are being calculated and displayed.</p>
<h4 id="apss-distribution-among-species-plot">APSS distribution among species plot</h4>
<p>This plot shows the APSS distribution among all the compared sample-pairs of each included species as a boxplot.<br />
The plot color can be changed using the color-picker widget.</p>
<p><strong>Including metadata:</strong></p>
<p>When checking the ‘Use metadata in plot’ checkbox, it is possible to select a feature, by which the comparisons can be divided into two categories: same / different.<br />
for example: if the selected feature is ‘country’, the two categories are ‘same country’ and ‘different country’.<br />
The features are derived from the uploaded metadata file and can be interactively selected from a drop-down menu.<br />
The colors of the same / different feature categories can be changed using the color-picker widgets.<br />
When using metadata, the P-values of the comparisons for the selected feature can be downloaded in addition to the APSS table.</p>
<h3 id="ani-multi-species-analysis">Multiple-species analysis of ANI results</h3>
<h4 id="ani-distribution-among-species-plot">ANI distribution among species plot</h4>
<p>This plot shows the ANI distribution among all the compared sample-pairs of each included species as a boxplot. It provides the same customization options as the APSS distribution among species plot, explained in details above.</p>

</main>
</body>
</html>